File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private TreeNode[] FlattenBinaryTree(TreeNode Root) {
        if (Root == null) return new TreeNode[2];
        TreeNode[] L = FlattenBinaryTree(Root.left), R = FlattenBinaryTree(Root.right);
        Root.left = null;
        if (L[0] != null) {
            Root.left = null;
            Root.right = L[0];
            L[1].right = R[0];
            return new TreeNode[]{Root, R[0] == null ? L[1] : R[1]};
        } else {
            Root.right = R[0];
            return new TreeNode[]{Root, R[0] == null ? Root : R[1]};
        }
    }
    public void flatten(TreeNode root) {
        FlattenBinaryTree(root);
    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        
        def helper(root):
            if root is None:
                return [None, None]
            L = helper(root.left)
            R = helper(root.right)
            root.left = None
            if L[0] is None:
                return [root, R[1] if R[0] is not None else root]
            root.right = L[0]
            L[1].right = R[0]
            return [root, R[1] if R[0] is not None else L[1]]
        
        helper(root)
        